src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupLinearFunctional.java

```diff
@@ -0,0 +1,516 @@
+		for(int row = 0, offTmp = 0; row < tCol; row++, offTmp += tCol) {
+			final double alpha1 = nRows * _coefficents[0][row] + sumIndices * _coefficents[1][row];
+			final double alpha2 = sumIndices * _coefficents[0][row] + sumSquaredIndices * _coefficents[1][row];
+			final int offRet = _colIndexes[row] * numColumns;
+			for(int col = row; col < tCol; col++) {
+				ret[offRet + _colIndexes[col]] += alpha1 * _coefficents[0][col] + alpha2 * _coefficents[1][col];
+			}
+		}
+	}
+
+	@Override
+	public void leftMultByMatrixNoPreAgg(MatrixBlock matrix, MatrixBlock result, int rl, int ru, int cl, int cu) {
+		throw new DMLCompressionException("This method should never be called");
+	}
+
+	@Override
+	public void leftMultByAColGroup(AColGroup lhs, MatrixBlock result) {
+		if(lhs instanceof ColGroupEmpty) {
+			return;
+		}
+
+		MatrixBlock tmpRet = new MatrixBlock(lhs.getNumCols(), _colIndexes.length, 0);
+
+		if(lhs instanceof ColGroupUncompressed) {
+			ColGroupUncompressed lhsUC = (ColGroupUncompressed) lhs;
+			int numRowsLeft = lhsUC.getData().getNumRows();
+
+			double[] colSumsAndWeightedColSums = new double[2 * lhs.getNumCols()];
+			for(int j = 0, offTmp = 0; j < lhs.getNumCols(); j++, offTmp += 2) {
+				for(int i = 0; i < numRowsLeft; i++) {
+					colSumsAndWeightedColSums[offTmp] += lhs.getIdx(i, j);
+					colSumsAndWeightedColSums[offTmp + 1] += (i+1) * lhs.getIdx(i, j);
+				}
+			}
+
+			MatrixBlock sumMatrix = new MatrixBlock(lhs.getNumCols(), 2, colSumsAndWeightedColSums);
+			MatrixBlock coefficientMatrix = DataConverter.convertToMatrixBlock(_coefficents);
+
+			LibMatrixMult.matrixMult(sumMatrix, coefficientMatrix, tmpRet);
+		} else if(lhs instanceof ColGroupLinearFunctional) {
+			ColGroupLinearFunctional lhsLF = (ColGroupLinearFunctional) lhs;
+
+			final double sumIndices = _numRows * (_numRows + 1)/2.0;
+			final double sumSquaredIndices = _numRows * (_numRows + 1) * (2*_numRows + 1)/6.0;
+
+			MatrixBlock weightMatrix = new MatrixBlock(2, 2, new double[] {_numRows, sumIndices, sumIndices, sumSquaredIndices});
+			MatrixBlock coefficientMatrixLhs = DataConverter.convertToMatrixBlock(lhsLF._coefficents);
+			MatrixBlock coefficientMatrixRhs = DataConverter.convertToMatrixBlock(_coefficents);
+
+			coefficientMatrixLhs = LibMatrixReorg.transposeInPlace(coefficientMatrixLhs,
+				InfrastructureAnalyzer.getLocalParallelism());
+
+			// We simply compute a matrix multiplication chain in coefficient space, i.e.,
+			// 					t(L) %*% R = t(coeff(L)) %*% W %*% coeff(R)
+			// where W is a weight matrix capturing the size of the shared dimension (weightMatrix above)
+			// and coeff(X) denotes the 2 x n matrix of the m x n matrix X.
+			MatrixBlock tmp = new MatrixBlock(lhs.getNumCols(), 2, false);
+			LibMatrixMult.matrixMult(coefficientMatrixLhs, weightMatrix, tmp);
+			LibMatrixMult.matrixMult(tmp, coefficientMatrixRhs, tmpRet);
+		} else if(lhs instanceof APreAgg) {
+			// TODO: implement
+			throw new NotImplementedException();
```

Comment:
i think the best design for this would be some kind of iterator for the functional group that gives double[] line by line that can then be aggregated.

You do not need (i suggest not to) to implement this now